### DISCUSSION TITLE
Calculation of Taylor factor - dependence on magnitude of strain?

### DISCUSSION THREAD

**Discussion by JustinYCheng** [mtex-toolbox/mtex#747] [Ask Anything]
When I use calcTaylor to calculate the Taylor factor, it seems that there is a dependence of the calculated Taylor factor on the magnitude of the strain tensor used for this calculation. For example if I use calcTaylor in this manner:

```matlab
% oriList - a list of orientations that I want to calculate the Taylor factor for
% eps - the strain tensor that I use
% sS - a single slip system that I symmetrise to obtain the set of slip systems that will be used in the computation
% M - my Taylor factors for each orientation
[M,~,~] = calcTaylor(inv(oriList)*eps,sS.symmetrise);
```
So if I define my strain tensor as:

```matlab
eps = strainTensor(diag(2*[-.5 -.5 1]))
```

The value of M (or the Taylor factor) obtained using calcTaylor will be 2 times the value of the factor that I get using:

```
eps = strainTensor(diag([-.5 -.5 1]))
```

To my understanding, the Taylor factor is not supposed to act like this - only the orientation/geometry of the strain is supposed affect the Taylor factor and not the magnitude. For example, both `eps` defined above specify uniaxial tension. The Taylor factor is supposed to have a constant value, regardless of the magnitude of strain in this case. Is there a reason calcTaylor behaves differently?

**Comment by kilir** [comment]
Hi,
as far as I understand the definition of the Taylor factor, you should normalize your tensors.
Cheers,
Rüdiger

**Reply by Mustafa-Rifat** [reply]
Hi Rüdiger,

I am currently working with Taylor Factor too. I have got my strain tensor from a finite element model. It goes like this:
`epsilon = strainTensor([0.00958486 0.00971106 0.043623; 0.00971106 0.0278111 0.0199594; 0.043623 0.0199594 0.037396],'rank',2);`
I am observing the same thing Justin mentioned. If I double the values of strain tensor, the M values gets doubled too. I have 2 queries regarding this.
1. How should I normalize the strain tensor. Pardon me if this is too naive to ask. Let's say, can I do it by dividing each of the elements by the maximum element value? 
2. Also, I am getting M values lesser than 1 now. Can M values be lesser than one? My material is FCC. The code I am using:
```
epsilon = strainTensor([0.00958486 0.00971106 0.043623; 0.00971106 0.0278111 0.0199594; 0.043623 0.0199594 0.037396],'rank',2);
sS = symmetrise(slipSystem.fcc(grains.CS));
[M,b,W] = calcTaylor(inv(grains.meanOrientation)*epsilon,sS);
figure
plot(grains,M)
mtexColorMap white2black
```
Thanks a lot!

**Reply by Mustafa-Rifat** [reply]
Just a follow up. I could normalize the strain tensor using `normalize(epsilon.M)`. Now I am getting M values greater than 1 (avg. is 3.04, whether previously it was 0.21). However, I will appreciate your thought on this.

**Reply by JustinYCheng** [reply]
Hi Mustafa, those values sound a lot more reasonable - pg. 62 of "The Mechanics of Crystals and Textured Polycrystals" by Hosford has a figure that shows that M should range from 2.449 to 3.674 for axisymmetric flow.

**Reply by JustinYCheng** [reply]
Hold on - I don't think this issue is as simple as it looks. When I normalize my eps from above, I get a strain tensor that doesn't resemble a state of uniaxial tension:
```matlab
-1.1547    0.5774   -0.5774
 0.5774   -1.1547   -0.5774
 0.5774    0.5774    1.1547
```
I think the issue here is conceptual - the Taylor factor is only meaningful in cases where you're trying to relate strains from multiple slip systems to an externally applied flow stress (or strain state). That means you have to have a component of maximum stress - you have to be in the principal frame of reference. If you are in this frame, then the normalization criterion is simple: the maximum principal strain has to have a magnitude of 1. 

This matches the definition of the Taylor factor:

![image](https://user-images.githubusercontent.com/66323610/106933730-dce99300-66de-11eb-9fb8-b5459824c976.png)

Where if we just calculate this discretely:

![image](https://user-images.githubusercontent.com/66323610/106933858-060a2380-66df-11eb-84c2-48f7cf25ba7e.png)

The shear strain ![image](https://user-images.githubusercontent.com/66323610/106933978-2fc34a80-66df-11eb-9102-27d0354856cd.png) (or more accurately sum of shear strains) would be what is calculated in `calcTaylor` (it just sums up shear strains on the minimum work-required slip systems without dividing by anything) if your max principal strain ![image](https://user-images.githubusercontent.com/66323610/106934174-6d27d800-66df-11eb-80b7-67371aa27c7c.png) is equal to 1.

So this means:
1. Apologies to @Mustafa-Rifat, you'd have to do some analysis to see what exactly the Taylor factor you get means physically - you're not in the principal frame of reference
2. @kilir - wouldn't a more appropriate normalization condition be to divide the strain tensor used for Taylor factor calculation by the value of its maximum principal strain, assuming the given strain tensor is given in terms of the principal frame of reference?

**Reply by Mustafa-Rifat** [reply]
Hello Justin,

I took the `eps = strainTensor(diag([-.5 -.5 1]))` as you defined and did `normalize(eps.M)`. I also did `normalize(2*eps.M)`. Please see the results attached. 
Also, my epsilon values are taken from FEA model of an element. This is a model of a nanoscratch test, which is not uniaxial. I apologize but I did not understand what you meant by 'you're not in the principal frame of reference'. 

![maxresdefault-3-1024x576](https://user-images.githubusercontent.com/60717121/106937539-e1687880-66eb-11eb-9434-6e8b91025104.jpg)

What I understand from the code is to apply `calcTaylor(eps,sS)`, the eps and sS both should be in crystal reference frame (Link: https://mtex-toolbox.github.io/strainTensor.calcTaylor.html). Now the strain tensor I get from the FEA is for specimen reference frame. The` inv(grains.meanOrientation)*epsilon` inside `calcTaylor(inv(grains.meanOrientation)*epsilon,sS)` expression takes the epsilon from specimen reference frame to crystal reference frame and that is what we needed to do. Please correct me if I am wrong. May be I am thinking it too simply.

**Reply by Mustafa-Rifat** [reply]
Hi Justin,

Is there any update on this work that you were doing?

**Reply by farshaaad** [reply]
> Hi Justin,
> 
> Is there any update on this work that you were doing?

@Mustafa-Rifat 
Hello Mustafa,
I have encountered with the same issue. Did you manage to solve your problem?

**Reply by Mustafa-Rifat** [reply]
This is how I finally did it:
strain_FEA=[0.0246076 -0.133807 0.302012; -0.133807 0.0540857 0.157968; 0.302012 0.157968 -0.0765761];%strain tensor from FEA model
normalized_strain_FEA=strain_FEA./norm(strain_FEA,'fro');%normalize the strain tensor
epsilon = strainTensor(normalized_strain_FEA);
sS = symmetrise(slipSystem.fcc(grains.CS));
%the inv transforms the strain tensor from specimen coordinate to crystal coordinate system,sS is already in crystal coordinate system
[M,b,W] = calcTaylor(inv(grains.meanOrientation)*epsilon,sS);
Let me know if you have any further queries.

**Reply by farshaaad** [reply]
@Mustafa-Rifat 
Thank you so much for your reply Rifat. I was working on it. Now I got it. Also, I have a question about the b which is the coefficients of the active slip systems. How can we define the active slip systems using b? There should be 5 active slip systems and each of them has a slip direction vector and normal to slip plane therefore it should be 30 values, but b presents 24 values. Am I missing something?

**Comment by Mustafa-Rifat** [comment]
To find the most active slip system, I would use Schmid Factor. The slip
system with maximum Schmid factor is the most active one.

**Reply by farshaaad** [reply]
So according to Taylor's model, there should be 5 active slip systems. Then you will need to find 5 slips systems which have the highest Schmid factor. However, in the Taylor model, it is possible to have more than one combination of 5 active slip systems. To solve this ambiguity different criteria have been suggested to choose one combination of the slip systems among the possible solutions. I do not know what criteria MTEX code use, but by using Schimd factor I do not know how we can define the active slip systems related to the calculated Taylor factor. Please let me know if I am missing something.

**Comment by ralfHielscher** [comment]
Finally, the function ```calcTaylor``` has been modified to return the Taylor factor with respect to the normalized strain tensor. See 5c0a5875204a57b486c41f4b303183f892edb10a.

Ralf.
